src/stdlib_stats_distribution_uniform.fypp

```diff
@@ -0,0 +1,484 @@
+    !!
+    !! Get uniform distribution cumulative distribution function (cdf) for integer, real and complex variables
+    !! ([Specification](../page/specs/stdlib_stats_distribution_uniform.html#
+    !! description))
+    !!
+        #:for k1, t1 in ALL_KINDS_TYPES
+        module procedure unif_dist_cdf_${t1[0]}$${k1}$
+        #:endfor
+    end interface uniform_distribution_cdf
+
+    interface shuffle
+    !! Version experimental
+    !!
+    !! Fisher-Yates shuffle algorithm for a rank one array of integer, real and complex variables
+    !! ([Specification](../page/specs/stdlib_stats_distribution_uniform.html#
+    !! description))
+    !!
+        #:for k1, t1 in ALL_KINDS_TYPES
+        module procedure shuffle_${t1[0]}$${k1}$
+        #:endfor
+    end interface shuffle
+
+
+    contains
+
+    #:for k1, t1 in INT_KINDS_TYPES
+    impure elemental function unif_dist_rvs_1_${t1[0]}$${k1}$(scale) result(res)
+    ! Uniformly distributed integer in [0, scale]
+    ! Bitmask with rejection
+    ! https://www.pcg-random.org/posts/bounded-rands.html
+    !
+    ! Fortran 90 translated from c by Jim-215-fisher
+        ${t1}$, intent(in) :: scale
+        ${t1}$ ::  res, u, mask, n
+        integer :: zeros, bits_left, bits
+
+        n = scale
+        if(n == 0_${k1}$) call error_stop("Error: Uniform distribution scale"   &
+                          //" parameter must be non-zero")
+        zeros = leadz(n)
+        bits = bit_size(n) - zeros
+        mask = shiftr(not(0_${k1}$), zeros)
+        L1 : do
+            u = dist_rand(n)
+            res = iand(u, mask)
+            if(res <= n) exit L1
+            bits_left = zeros
+            L2 : do
+                if(bits_left < bits) exit L2
+                u = shiftr(u, bits)
+                res = iand(u, mask)
+                if(res <= n) exit L1
+                bits_left = bits_left - bits
+            end do L2
+        end do L1
+        return
+    end function unif_dist_rvs_1_${t1[0]}$${k1}$
+
+    #:endfor
+
+    #:for k1, t1 in INT_KINDS_TYPES
+    impure elemental function unif_dist_rvs_${t1[0]}$${k1}$(loc, scale)         &
+                     result( res )
+    ! Uniformly distributed integer in [loc, loc + scale]
+    !
+        ${t1}$, intent(in) :: loc, scale
+        ${t1}$  ::  res
+
+        if(scale == 0_${k1}$) call error_stop("Error: Uniform distribution"     &
+                                    //" scale parameter must be non-zero")
+        res = loc + unif_dist_rvs_1_${t1[0]}$${k1}$(scale)
+        return
+    end function unif_dist_rvs_${t1[0]}$${k1}$
+
+    #:endfor
+
+    #:for k1, t1 in REAL_KINDS_TYPES
+    impure elemental function unif_dist_rvs_0_${t1[0]}$${k1}$( ) result(res)
+    ! Uniformly distributed float in [0,1]
+    ! Based on the paper by Frederic Goualard, "Generating Random Floating-
+    ! Point Numbers By Dividing Integers: a Case Study", Proceedings of
+    ! ICCS 2020, June 20202, Amsterdam, Netherlands
+    !
```

Comment:
Note that Goualard mentions dividing integers by a Mersenne number to make a floating-point number to make a point that doing so can lead to non-uniformity issues, and not necessarily to show that libraries should generate random floating-point numbers this way.  He makes the same point with dividing integers by a power of 2.

Also, there are ways to implement uniform floating-point number generation that covers all numbers the `real` type covers in [0, 1], but in general they are far from trivial which is why I don't demand it here.  Examples include the Rademacher Floating-Point Library by @christophe-conrads as well as an [algorithm I give](https://peteroupc.github.io/randomfunc.html#For_Floating_Point_Number_Formats).

Reply:
Not quite understand your point. Yes, dividing Mersenne integer leads to non-uniformity. Multiplying the same number reduces the issue significantly.

Reply:
My point is indeed to reduce non-uniformity, especially since this code is intended to form part of the standard library and thus be used by thousands of applications with different requirements, and some of them may indeed care about the statistical quality of pseudorandom numbers.  In this sense, of the methods that involve multiplying or dividing an integer by a constant to create a binary64, the "best" one is to multiply a binary64 integer in [1, 53^2) (which is representable in binary64) by the binary64 constant 1/53^2.  This is "better" than multiplying an integer by a Mersenne number or its inverse, because it "better" ensures the binary64 values are evenly spaced in [0, 1] as the result of the multiplication is representable in binary64. This difference is illustrated if we multiply 5201 by either 1/53^2 or 1/53^2-1, then convert the result to binary64 and back. (Notice that I put "best" and "better" in quotes here, in part because, in my opinion, floating-point random numbers should be avoided whenever random integers or random rational numbers suffice.)

Reply:
Based on Goualard paper, the best  one is a binary64 integer in [1, 2^53) multiplying a float of 1/(2^53-1) as shown in Fig 7. in the paper. This is exactly what I implemented in the module. Dividing binary64 integer by 2^53  leads to more non-uniformity as GNU Fortran 9.2.0 does.